### PR TITLE
Alternative fix to issue 17548 - [REG2.072.0] Forward reference error with scope function parameters

### PR DESCRIPTION
**Pull request by Syniurge**
Previous PR: #6934

This one line fix ensures that if a `StructDeclaration.semantic()` call has to defer itself, `semanticRun` is set to `PASSsemantic`, because it may already have been set to `PASSsemanticdone` if a second `StructDeclaration.semantic()` call on the same struct happened while semantic'ing its members (see how in the Bugzilla issue) and then the deferred call would immediately return.

### CODE REVIEW

**Comment by CyberShadow** [comment]
(sorry about that, misclicked)

**Comment by Syniurge** [comment]
I have a question about this code in `StructDeclaration.semantic`:
```d
    else if (symtab && !scx)
    {
        semanticRun = PASSsemanticdone;
        return;
    }
```
What's the rationale? Is it assuming that this may only happen if there was a previous `StructDeclaration.semantic` call which failed with:
```d
        if (!determineFields())
        {
            assert(type.ty == Terror);
            sc2.pop();
            return;
        }
```
which is why it sets `semanticRun` to `PASSsemanticdone`?

If yes the assumption's wrong, and a proper fix would probably be to get rid of that assignment?

**Comment by Syniurge** [comment]
So this works too:
```d
         else if (symtab && !scx)
 -        {
 -            semanticRun = PASSsemanticdone;
              return;
 -        }

         (...)

         if (!determineFields())
          {
              assert(type.ty == Terror);
              sc2.pop();
 +            semanticRun = PASSsemanticdone;
              return;
          }
```
and I think this is the proper fix.

Why isn't the auto-tester running though?

**Comment by CyberShadow** [comment]
@braddr Here's another instance of the autotester apparently ignoring a PR.

**Comment by braddr** [comment]
In this case, it's easy.. new user, not approved yet for builds to be processed.  Just look here:
https://auto-tester.puremagic.com/pulls.ghtml?projectid=14

**Comment by CyberShadow** [comment]
Aah, forgot about that, thanks!

**Comment by UplinkCoder** [comment]
@WalterBright this looks okay at a glance.

**Comment by Syniurge** [comment]
Ping?

File: src/ddmd/dstruct.d [hunk 1]
```diff
@@ -323,10 +323,8 @@ extern (C++) class StructDeclaration : AggregateDeclaration
             userAttribDecl = sc.userAttribDecl;
         }
         else if (symtab && !scx)
-        {
-            semanticRun = PASSsemanticdone;
```
Comment (MartinNowak) [inline src/ddmd/dstruct.d]: Looks good, so the problem was that we set this to done here, even though we might be in a recursive call to StructDeclaration.semantic, with the outer semantic still ongoing on it's members, hence providing wrong semanticRun state info for any further members.

**Review by MartinNowak — APPROVED** [review]

**Pull request merged**